libraries/Microsoft.Bot.Connector/Authentication/AdalAuthenticator.cs

```diff
@@ -0,0 +1,188 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Connector.Authentication
+{
+    public class ThrottleException : Exception
+    {
+        public RetryParams RetryParams { get; set; }
+    }
+    
+    public class AdalAuthenticator 
+    {
+        // Our ADAL context. Acquires tokens and manages token caching for us.
+        private readonly AuthenticationContext authContext;
+
+        // Semaphore to control concurrency while refreshing tokens from ADAL.
+        // Whenever a token expires, we want only one request to retrieve a token. 
+        // Cached requests take less than 0.1 millisecond to resolve, so the semaphore doesn't hurt performance under load tests
+        // unless we have more than 10,000 requests per second, but in that case other things would break first.
+        private static Semaphore tokenRefreshSemaphore = new Semaphore(1, 1);
+        private static readonly TimeSpan SemaphoreTimeout = TimeSpan.FromSeconds(10);
+
+        // Depending on the responses we get from the service, we update a shared retry policy with the RetryAfter header
+        // from the HTTP 429 we receive.
+        // When everything seems to be OK, this retry policy will be empty.
+        // The reason for this is that if a request gets throttled, even if we wait to retry that, another thread will try again right away.
+        // With the shared retry policy, if a request gets throttled, we know that other threads have to wait as well.
+        // This variable is guarded by the authContextSemaphore semphore. Don't modify it outside of the semaphore scope.
+        private static volatile RetryParams currentRetryPolicy;
+
+        private readonly ClientCredential clientCredential;
+        private readonly IOAuthConfiguration oAuthConfig;
+
+        private const string msalTemporarilyUnavailable = "temporarily_unavailable";
+
+        public AdalAuthenticator(ClientCredential clientCredential, IOAuthConfiguration oAuthConfig)
+        {
+            this.oAuthConfig = oAuthConfig;
+            this.clientCredential = clientCredential ?? throw new ArgumentNullException(nameof(clientCredential));
+            this.authContext = new AuthenticationContext(oAuthConfig.Authority);
+        }
+
+        public async Task<AuthenticationResult> GetTokenAsync(bool forceRefresh = false)
+        {
+            return await Retry.Run(
+                task: () => AcquireTokenAsync(forceRefresh),
+                retryExceptionHandler: (ex, ct) => HandleAdalException(ex, ct)).ConfigureAwait(false);
+        }
+
+        private async Task<AuthenticationResult> AcquireTokenAsync(bool forceRefresh = false)
+        {
+            bool acquired = false;
+
+            if (forceRefresh)
+            {
+                authContext.TokenCache.Clear();
+            }
+
+            try
+            {
+                // The ADAL client team recommends limiting concurrency of calls. When the Token is in cache there is never 
```

Comment:
Sure! The advice came from principal members of the adal .net team, i.e. the team that wrote the code you pointed. Happy to share email threads if you want to see the proof. They also mentioned that MSAL (the soon to be GA revamped version) would natively have concurrency control so as callers we don't have to be so careful and would be more suitable for server applications out of the box, which made me happy.

Regarding the semaphore making it worse, here is the experimental data I collected. All collected in a default azure web app instance + azure cloud load tests. 

- If I don't do any concurrency control, in a load test when a token expires after ~200 requests each call takes more than 20 seconds and soon after all requests time out. 

- With this implementation there is not a single failure in load tests when we refresh the token at pretty high requests per second and the cruising load we can take is so high that usually other parts of the infrastructure fail before this. 

I'm not saying it is perfect, I wish I didn't have to do all this. With the experimental results I mention above, I don't quite understand the foundation of your comment that the semaphore is really just making it worse. I do dislike adding any kind of unnecessary concurrency control though.

Regarding the last paragraph, the implementation I'm replacing is a custom implementation written by partner MSFT FTEs and proved to not be robust at all, so I'd not lightly say that it is not really all that hard.

Thanks @drub0y for scrutinizing this so carefully, it's a luxury to have folks review PRs with so much detail!

Reply:
> Happy to share email threads if you want to see the proof

Oh, no, I totally believe you of course. I just wonder why they are giving this advice. Do they not realize their `TokenCache` implementation is already locking basically _everything_? The advice kind of makes no sense if you go look at that code... do you agree? Am I missing something? For the record, [here's the PR that added the locking](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/issues/308) and it sure seems the kind of fix that's like "i'll lock everything and we won't have any more concurrency bugs!"

> With the experimental results I mention above, I don't quite understand the foundation of your comment that the semaphore is really just making it worse. I do dislike adding any kind of unnecessary concurrency control though.

Sorry, what I mean is, the advice has us putting locking logic into the BBSDK's core path which, as it's currently written, is going to serialize every single request thread behind a single gate, right? Ok, yeah, there's a runtime overhead of entering the semaphore every time, but let's put that aside. I'm more thinking about the fact that forces locking logic into the BBSDK codebase which means it's something that BBSDK is now accountable for maintaining.

Now if I go file an issue with them about this and they eventually improve the locking in their implementation to make it as efficient as possible? Well, awesome, the entire world can benefit by just upgrading their package reference to the latest minver, except the BBSDK will still have a lock left that completely serializes every request and until that work is undone every bot continues to "suffer" from serialization of requests. If we can avoid putting locking logic into this code because it's clear the implementation is thread safe on its own, then we avoid that maintenance issue.

Finally, if nothing else in this timeframe, can I at least convince you to switch to a `SemaphoreSlim`? 😊

Reply:
I agree on their token cache logic and the fix, lol. However even that lock is over-serializing, we have to look the entire operation (check cache, refresh token if needed, update cache) which is what, if under high concurrency, simply starts failing after a few couple requests as I mentioned on my experimental data.

I would defnitely consider to a sempahore slim or readerwriterlockslim for this iteration, absolutely - granted that the load test results are equal or better than the current implementation. I actually started with a sempahore slim because it does make more sense but I saw better performance with the semaphore despite the obvious fact that semaphores are more expensive. Unfortunately I did was not able to understand the reason for that after a short analysis and then I just had to move on to other stuff.

I wholeheartedly agree with the fact that I hate adding this lock and serializing requests!!! Still, even though conceptually and even from a principles standpoint it is a nightmare for me, I need to guarantee that this works at high load for our big customers with high traffic such as vodaphone and telefonica, and right now this is an implementation that guarantees that. Even though we are serializing, 99.999999 of the time it comes from the cache in <<1 millisecond so there is no contention on the lock.

I am in sync with the ADAL.NET team very frequently, i.e. this week they announced me personally about some news on GA for MSAL.NET, the new library and had a conversation with their lead PM. As soon as they do improve the locking on their implementation to make it more efficient and server friendly as well, we can just remove the locking and have then maintain the concurrency control, which is great!

Hope that makes sense Drew! Again, thanks for your attention and thoughtful comments :)